fix(vertex_ai/gemini): improve chunk parsing for streaming responses

miraclebakelaser opened the pull request:
This PR fixes bug #8143 by updating the vertex_ai streaming chunk parsing logic to remove only a leading `data:` prefix.



## Relevant issues
#8143 


## Type

🐛 Bug Fix

## Changes

```diff
    def _common_chunk_parsing_logic(self, chunk: str) -> GenericStreamingChunk:
        try:
-            chunk = chunk.replace("data:", "")
+            chunk = chunk.strip()
+            if chunk.startswith("data:"):
+                chunk = chunk[len("data:"):].strip()
             if len(chunk) > 0:
                 """
                 Check if initial chunk valid json
                 - if partial json -> enter accumulated json logic
                 - if valid - continue
                 """
                 if self.chunk_type == "valid_json":
                     return self.handle_valid_json_chunk(chunk=chunk)
                 elif self.chunk_type == "accumulated_json":
                     return self.handle_accumulated_json_chunk(chunk=chunk)
```

miraclebakelaser committed:
fix(vertex_ai/gemini): improve chunk parsing for streaming responses

Fix incorrect removal of "data:" strings within response content.

ishaan-jaff commented on litellm/llms/vertex_ai/gemini/vertex_and_google_ai_studio_gemini.py:
```diff
@@ -1431,7 +1431,9 @@ def handle_accumulated_json_chunk(self, chunk: str) -> GenericStreamingChunk:
-            chunk = chunk.replace("data:", "")
+            chunk = chunk.strip()
+            if chunk.startswith("data:"):
+                chunk = chunk[len("data:"):].strip()
```
`.strip()` Is this needed ? Wouldn't this remove an start / end empty space that actually came from llm ?

miraclebakelaser replied:
In the case where `self.chunk_type == "valid_json"`, the strip wouldn't have any effect on the llm message content itself (added testing for it). I updated the handling to be more precise.

ishaan-jaff reviewed (changes requested):
please a add a test and / or unit test

miraclebakelaser committed:
test(vertex_ai/gemini): add tests for SSE stream chunk parsing fix

Enhance test coverage for ModelResponseIterator's chunk parsing logic, focusing on preserving 'data:' prefixes and surrounding spaces in Vertex AI and Google Studio Gemini streaming responses

miraclebakelaser committed:
More precise handling of 'data: ' prefix

miraclebakelaser commented:
> please add a test and / or unit test

Added tests that cover cases where `self.chunk_type == "valid_json"`. How are chunks formatted when `self.chunk_type == "accumulated_json"`? 

I understand that in the case of `self.chunk_type == "valid_json"`, the messages arrive in event stream format and are chunked by message:

chunk 1: 
`data: {"candidates": [{"content": {"role": "model","parts": [{"text": "data"}]}}],"usageMetadata": {},"modelVersion": "gemini-2.0-flash-exp","createTime": "2025-01-01T00:00:00.000000Z","responseId": "12345"}`

chunk 2:
`data: {"candidates": [{"content": {"role": "model","parts": [{"text": ": line "}]}}],"modelVersion": "gemini-2.0-flash-exp","createTime": "2025-01-01T00:00:00.000000Z","responseId": "12345"}`

chunk 3:
`data: {"candidates": [{"content": {"role": "model","parts": [{"text": "1\\ndata: line 2\\ndata: line 3\\ndata:"}]}}],"modelVersion": "gemini-2.0-flash-exp","createTime": "2025-01-01T00:00:00.000000Z","responseId": "12345"}`

chunk 4: 
`data: {"candidates": [{"content": {"role": "model","parts": [{"text": " line 4\\n"}]},"finishReason": "STOP"}],"usageMetadata": {"promptTokenCount": 42,"candidatesTokenCount": 24,"totalTokenCount": 66,"promptTokensDetails": [{"modality": "TEXT","tokenCount": 42}],"candidatesTokensDetails": [{"modality": "TEXT","tokenCount": 24}]},"modelVersion": "gemini-2.0-flash-exp","createTime": "2025-01-01T00:00:00.000000Z","responseId": "12345"}`

When `self.chunk_type == "accumulated_json"`, do chunks have the same format as above (i.e. `data: ...`), or are they something like a json string split into pieces like the following:

chunk 1:
`{"candidates": [{"content": {"role": "model","parts`

chunk 2:
`": [{"text": "```"}]}}],"usageMetadata": {},"modelVersion": "gemini-1.5-flash-001","createTime":`


I haven't encountered a situation where the code goes down this path so your input would help me add a test for that case too.

fengjiajie commented:
@miraclebakelaser Thanks for fixing this issue! It looks like `handle_accumulated_json_chunk` in this same file might have a similar problem? I think it might need the same change. Also, you might want to mark ishaan-jaff's previous review requests as resolved – that way, they'll be notified to take another look.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=8401) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=8401) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=8401) it.</sub>
